src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -369,14 +378,20 @@ String getFieldType(String fieldName, Map<String, Object> propertiesAsMap, Index
         String fieldType = getFieldTypeFromCache(fieldName, index);
 
         if (fieldType != null) {
+            cacheHitCount += 1;
             return fieldType;
+        } else {
+            cacheMissCount += 1;
         }
 
         // Retrieve field type from mapping and cache it if found
         fieldType = getFieldTypeFromProperties(fieldName, propertiesAsMap);
 
         // Cache field type or NO_FIELD_TYPE_VALUE if not found
-        indicesFieldTypeCache.getOrInitialize(index).putIfAbsent(fieldName, fieldType != null ? fieldType : NO_FIELD_TYPE_VALUE);
+        fieldType = fieldType != null ? fieldType : NO_FIELD_TYPE_VALUE;
+        if (indicesFieldTypeCache.getOrInitialize(index).putIfAbsent(fieldName, fieldType)) {
+            indicesFieldTypeCache.incrementWeight(fieldName, fieldType);
```

Comment:
Isn't the weight incremented twice, during `putIfAbsent` and also during `incrementWeight`?

Reply:
There is a weight for the overall cache `IndicesFieldTypeCache` and individual weights for the per-index maps `IndexFieldMap`